pkg/webhook/webhooks.go

```diff
@@ -0,0 +1,280 @@
+	//var pvc corev1.PersistentVolumeClaim
+	//err := json.Unmarshal(req.Object.Raw, &pvc)
+	//if err != nil {
+	//	glog.Errorf("Could not unmarshal raw object: %v, %v", err, req.Object.Raw)
+	//	status.Allowed = false
+	//	status.Result = &metav1.Status{
+	//		Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+	//		Message: err.Error(),
+	//	}
+	//	return status
+	//}
+
+	// fetch the pvc specifications
+	pvc, err := wh.KubeClient.CoreV1().PersistentVolumeClaims(req.Namespace).Get(req.Name, metav1.GetOptions{})
+	if err != nil {
+		status.Allowed = false
+		status.Result = &metav1.Status{
+			Status: metav1.StatusFailure, Code: http.StatusBadRequest, Reason: metav1.StatusReasonBadRequest,
+			Message: err.Error(),
+		}
+		return status
+	}
+
+	if !validationRequired(ignoredNamespaces, &pvc.ObjectMeta) {
+		glog.Infof("Skipping validation for %s/%s due to policy check", pvc.Namespace, pvc.Name)
+		return &v1beta1.AdmissionResponse{
+			Allowed: true,
+		}
+	}
+
+	// get the all CStorVolumes resources in openebs namespace(default namespace
+	// for CStorvolume resource) to check the source-volume annotation to find
+	// out if there is any clone volume exists.
+	// if source-volume annotation matches with name of PVC, failed the pvc
+	// deletion operation.
+	status.Allowed = true
+	cStorVolumes, _ := wh.Clientset.OpenebsV1alpha1().CStorVolumes("openebs").List(metav1.ListOptions{})
+	for _, cstorvolume := range cStorVolumes.Items {
+		if cstorvolume.Annotations["openebs.io/source-volume"] == pvc.Spec.VolumeName {
+			status.Allowed = false
+			status.Result = &metav1.Status{
+				Status: metav1.StatusFailure, Code: http.StatusForbidden, Reason: "the PVC has cloned volume created",
+				Message: fmt.Sprintf("pvc %q has cloned volume PV: %q", pvc.Name, cstorvolume.Name),
+			}
+		}
+	}
+	return status
+}
+
+// Serve method for webhook server, handles http requests for webhooks
+func (wh *Webhook) Serve(w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	if r.Body != nil {
+		if data, err := ioutil.ReadAll(r.Body); err == nil {
+			body = data
+		}
+	}
+	if len(body) == 0 {
+		glog.Error("empty body")
+		http.Error(w, "empty body", http.StatusBadRequest)
+		return
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		glog.Errorf("Content-Type=%s, expect application/json", contentType)
+		http.Error(w, "invalid Content-Type, expect `application/json`", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	var admissionResponse *v1beta1.AdmissionResponse
+	ar := v1beta1.AdmissionReview{}
+	if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+		glog.Errorf("Can't decode body: %v", err)
+		admissionResponse = &v1beta1.AdmissionResponse{
+			Result: &metav1.Status{
+				Message: err.Error(),
+			},
+		}
+	} else {
+		if r.URL.Path == "/validate" {
```

Comment:
Does this URL follow REST standard?

Reply:
It's just to append this to the kubernetes api-extension endpoint...